tests/test_type_caster_bare_interface.cpp

```diff
@@ -0,0 +1,215 @@
+// Systematically exercises the detail::type_caster<> interface. This is going a step in the
+// direction of an integration test, to ensure multiple components of pybind11 work together
+// correctly. It is also useful to show the type_caster<> interface virtually clutter-free.
+
+#include "pybind11_tests.h"
+
+#include <memory>
+
+namespace pybind11_tests {
+namespace type_caster_bare_interface {
+
+struct mpty {};
```

Comment:
I have a hard time reading everything since 1) I have no idea how to pronounce/read this, 2) my brain's expecting capitals for "custom"/user-side classes in the test. Do you mind just turning (or I can also do it, if you don't have time) this into `Foo` or `CustomClass` or `Value` or so?